Recipe packs (#109)
radius-project/design-notes: recipe/2025-08-recipe-packs.md

Reviewer: Can you give more detail on the scenario in which we would want the deployment to fail? If I am deploying another resource using the same recipe pack, that should succeed. If I am re-deploying a resource that is already deployed using the same recipe pack, that should succeed too. What is the scenario that should fail?

Author: Upadted since we chose to not bring in versioning for the 1st iteration.

## Terms and definitions

| Term     | Definition                                                                                                                                                                                                 |
| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Recipe | IaC templates that operators register on a Radius Environment |
| Recipe Pack| A collection of recipes that can be managed as an entity |
| RRT | [Radius Resource Type](https://docs.radapp.io/guides/author-apps/custom/overview/)|

## Objectives

> **Issue Reference:** 

### Goals

Provide Recipe Packs as a Radius feature to bundle multiple recipes as a single managable entity into a Radius Environment (e.g., a pack for ACI that includes all necessary recipes).

### Non goals

Recipe Packs would bundle together Recipes, as we understand them today. We do not cover recipe versioning / other recipe specific enhancements in this design. 

> ### User scenarios (optional)

Reviewer 2: One more scenario I suggest adding is the ability to share Recipe Packs across environments or even organizations. e.g. Azure or AWS might create Recipes and bundle them into a Recipe Pack that can be used off the shelf, compute platform specific packs (e.g. Kubernetes, ACI) that ship with Radius by default

#### Registering several recipes to an environment

As an operator I am responsible for creating Radius Environments using which developers can deploy their applications. As part of creating the environment, I manually link recipes one by one using `rad recipe register` or by updating the environment definition. This can be error prone when there are many recipes and environments.  Radius should provide a way to bulk register( and manage) recipes.

> #### Registering recipes to multiple environment

Reviewer 2: ```suggestion
#### Registering the same set of recipes to multiple environments
```

As an operator I am responsible for creating Radius Environments using which developers can deploy their applications. As part of creating the environment, I manually link recipes one by one using `rad recipe register` or by updating the environment definition. I having 100s of environment which mostly use the same recipes. Piecing the same recipes together for each environment feels like rework.

## Design

### Design Overview

We choose modelling Recipe Pack as a first class Radius Resource Type. [Alternatives considered](#alternatives-considered) section details some other possibilities.

Pros:

- Solves the requirement for bulk registering recipes using single command with a one time effort of creating the recipe pack resource
- As first class resource, recipe packs would be displayed in app graphs. They can also have their own lifecycle and rbac independant of environments. 
- Helps reduce the size of environment resource, which could reach serization limits with tons of recipes. 
- Helps reduce overall size of Radius datastore, since common recipe information could now be stored as a single resource instead of being duplicated across several environments.

> [Question]: would environment and recipe pack ever have different rbac?

Reviewer 2: I don't think Recipe Packs have RBAC? They should just inherit whatever RBAC is needed to drive its constituent Recipes, which in turn inherits RBAC from the Environment it's registered to?

Cons:

- This approach is a deviation from the current tooling approach for recipes. 

> - While this brings in many advantages, RRTs can have their schema modified using rad resource-type commands. We should find ways to prevent this from happening. In general, Radius.Core namespace would have resources whose schema should be non-editable so that Radius can work as expected. Some other resources that would fall in this category are environments and applications.

Reviewer 2: This is an interesting take - we'd have to make parts of the Recipe Pack schema non-editable while other parts of it must remain editable (e.g. the set of Recipes bundled into the Pack)

              recipes:
                type: object
                description: Map of resource types to their recipe configurations
                additionalProperties:
                  type: object
                  properties:
                    recipeKind:
                      type: string
                      description: The type of recipe (e.g., terraform, bicep)
                      enum:
                        - terraform
                        - bicep
                    recipeLocation:
                      type: string
                      description: URL or path to the recipe source
                    parameters:
                      type: object
                      description: Parameters to pass to the recipe
                      additionalProperties:
                        type: any
                  required:
                    - recipeKind
                    - recipeLocation
            required:
              - name
              - recipes
```

Today the RRT schema does not allow `any` as a type for security reasons. We would have to remove that constraint in bicep tooling and radius so that we can have the recipe parameters as defined above.

> // Question: The recipe collection should be curly braces, is that OK (feature spec has []) ? OR we keep it array like below. Map could make it easier to look up using resource type.

Reviewer 2: I'm indifferent. If map works better, that's alright!

"recipes": [
    {
      "resourceType": "Radius.Compute/containers",
      "recipeKind": "terraform",
      "recipeLocation": "https://example.com/recipes/containers.zip"
    },
    {
      "resourceType": "Radius.Storage/volumes",
      "recipeKind": "terraform",
      "recipeLocation": "https://example.com/recipes/volumes.zip"
    }
  ]

### API design (if applicable)

> We should support CRUDL operations on recipe-pack resource. This should be fairly automatic since recipe packs would be registered as an RRT. Once we register the schema for `Radius.Core/recipePacks` resource type using `rad resource-type create`,  Dynamic RP should be able to dynamically look up the schema and perfom these operations. 

Reviewer 2: Question to ponder: should Recipe Packs resource not be modeled as an RRT? Since it's a Radius construct and we don't expect users to need to customize their Recipe Packs?

- Imperative provisioning prevents users from accidentally experimenting with the resource schema

Cons:

- This approach is a deviation from the current tooling approach for recipes. 
- Supporting recipe packs as a Applications RP type requires manual implementation of schema and API in contrast to modeling it as a dynamic resources. 

At a very high level, this design approach needs the below steps:

* Add support for Radius.Core/recipePacks resource in Applications RP 
  * Schema + API design and implementation
* Support `rad cli` commands that enable CRUDL operations on resources of type Radius.Core/recipePacks 
* Design and support Radius.Core/environment schema in Applications RP
* Support `rad cli` commands that enable managing Radius.Core/environment resources through CRUDL operations on this type of resource
* Support `rad cli` commands that enable registering recipe-packs to a `Radius.Core/environments` environment resource
* Enhance Dynamic RP, Applications RP and UCP to support the feature.

### Schema and API design

> As part of supporting Recipe Pack as a resource type, at a high-level, we define a recipePacks.tsp

Author: updated design doc to reflect the choice of app rp for radius.core.

/radius/deploy/Chart/templates/ucp/configmaps.yaml

```yaml
initialization:
  planes:
    - id: "/planes/radius/local"
      properties:
        resourceProviders:
          Applications.Core: "http://applications-rp.radius-system:5443"
          Applications.Dapr: "http://applications-rp.radius-system:5443"
          Applications.Datastores: "http://applications-rp.radius-system:5443"
          Applications.Messaging: "http://applications-rp.radius-system:5443"
          Microsoft.Resources: "http://bicep-de.radius-system:6443"
        kind: "UCPNative"
```

#### Applications RP

There are two prerequisites for implementing recipe packs:

1. Support new Radius.Core namespace. We would add Radius.Core namespace so that there is a transition time for users to move from Applications.Core/environments and recipes to Radius.Core/environment and Radius.Core/recipePacks. Eventually Applications.Core will be deprecated and removed.

2. Radius.Core/environments design and implementation.
   
Below changes are needed for supporting recipe packs as new feature:

> 1. Add schema /swagger changes to support the `Radius.Core/recipePacks` resource type ([typespec changes](#schema-and-api-design))

Reviewer 3: Let's discuss this.

```bicep
computeRecipePack.bicep:

resource computeRecipePack 'Radius.Core/recipePacks@2025-05-01-preview' = {
  name: 'computeRecipePack'
  description: "Recipe Pack for deploying to Kubernetes."
  properties: {
    recipes: {
      'Radius.Compute/containers': {
        recipeKind: 'terraform'
        recipeLocation: 'https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48'
        parameters: {
          allowPlatformOptions: true
        }
      }
      'Radius.Security/secrets': {
        recipeKind: 'terraform'
        recipeLocation: 'https://github.com/project-radius/resource-types-contrib.git//recipes/security/secrets/kubernetes?ref=v0.48'
      }
      'Radius.Storage/volumes': {
        recipeKind: 'terraform'
        recipeLocation: 'https://github.com/project-radius/resource-types-contrib.git//recipes/storage/volumes/kubernetes?ref=v0.48'
      }
    }
  }
}

rad deploy computeRecipePack.bicep
```

Note: rad recipe-pack create command could be added in future. For now, we are using rad deploy to create recipe packs. 

2. Show recipe pack details:

```

> $ rad recipe-pack show computeRecipePack

Reviewer 4: There needs to be some way to see Recipe parameters.

Reviewer 4: We should output parameters on `rad recipe-pack show`. We're outputing all properties except parameters.

Author: updated. Please review.

RESOURCE                TYPE                      GROUP     STATE
computeRecipePack       Radius.Core/recipePacks   default   Succeeded

RESOURCE TYPE                    RECIPE KIND     RECIPE VERSION      RECIPE LOCATION
Radius.Compute/containers        terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48
Radius.Security/secrets          terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/security/secrets?ref=v0.48
Radius.Storage/volumes           terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/storage/volumes?ref=v0.48
```

3. List recipe packs:

```

> $ rad recipe-pack list [optionally -e myenv]

Reviewer 4: Add a `--group` arguement.

Reviewer 4: The use cases is "As a platform engineer, I'm creating an Environment. I need to to determine which Recipe Packs to include in my Environment. I expect `rad recipe-pack list` to list all Recipe Packs across all Resource Groups, but I may need to filter to just a single Resource Group."  

We do not want a `rad recipe-pack list --environment`. If the platform engineer wanted to see the Recipe Packs for an Environment, the proper command is `rad environment show` which must include the Recipe Packs on that Environment.

> $ rad recipe-pack list [optionally -e myenv]

Reviewer 4: ```suggestion
$ rad recipe-pack list
```
We should not have a `--environment` filter. The `rad environment show` is the proper way to see Recipe Packs on an Environment.

Author: This lists all recipe packs in current scope (radius ersource group), right?

Reviewer 4: No. The proper command is `rad environment show`. This is specified in the [feature spec](https://github.com/radius-project/design-notes/blob/main/features/2025-06-compute-extensibility-feature-spec.md#user-story-1-as-a-platform-engineer-i-want-to-create-a-recipe-pack-that-bundles-multiple-recipes-for-core-resource-types-so-that-i-can-easily-register-and-manage-them-in-a-radius-environment).

RESOURCE                TYPE                      GROUP     STATE
computeRecipePack       Radius.Core/recipePacks   default   Succeeded
dataRecipePack          Radius.Core/recipePacks   default   Succeeded
```

> If no -e is specified, this command should list all recipe packs in current scope. 

Reviewer 4: Global, all groups. The result set would only be ~10s.

4. Delete recipe pack:

```
$ rad recipe-pack delete <recipe-pack-name>
```

> We could delete recipe-packs that are not referenced by any environment in any resource-group. This requires further thought and is similar to deletion of resources such as resource-type resource we have today.

Author: yes, delete is complex and needs its own deisgn. Simple solution is to not delte a recipe pack that is referenced by any env. I think, then "What if there are resources deployed using those recipe packs" would be a decision while deleting a env that has been used in deploying these resources. @lakshmimsft @willtsai @zachcasper for more thoughts.

> We could delete recipe-packs that are not referenced by any environment in any resource-group. This requires further thought and is similar to deletion of resources such as resource-type resource we have today.

Author: block the action if there are refernces and be explicit on the decision.

> We could delete recipe-packs that are not referenced by any environment in any resource-group. This requires further thought and is similar to deletion of resources such as resource-type resource we have today.

Author: huddle on the delete decision.

> 5. Register recipe pack to environment:
> 
> ```
> $ rad env register recipe-pack recipe-pack-name [-g group-id]
> ```
> The command should use the resource ID to identify the environment is `Radius.Core\environment` resource and only then allow recipe pack registration. If the namespace is `Applications.Core`, it should work as it does today.
> 
> We specify group resource id if the recipe-pack is in a different scope from the environment.

Reviewer 4: ```suggestion
5. Create an Environment with a Recipe Pack:

$ rad environment create myEnv

If Recipe Packs is not specified when creating the Environment, the `rad environment create` command sets default Recipe Packs (the same behavior as `rad init`).

$ rad environment create myEnv \
  --recipe-packs computeRecipePack

This command creates an environment with the `recipePacks[]` property populated with `computeRecipePack`.

$ rad environment update myEnv \
  --recipe-packs computeRecipePack

The command should use the resource ID to identify the environment is `Radius.Core\environment` resource and only then allow the `recipePack` property to be updated. If the namespace is `Applications.Core`, an error is provided.

$ rad environment update myEnv \
  --recipe-packs otherResourceGroup/computeRecipePack

If the Recipe Pack is in a different Resource Group than the Environment, the Resource Group is passed as a prefix to the Recipe Pack name.
```
Let's not use the term `register`. It's ambiguous and is only used for Recipes and Credentials today. Everything else is Create, Show, List, Delete.

Author: I thought we use register for pointers and create for creating a radius resource. For example we dont create recipes we just point to them in env.  But we create recipe pack resource. Please let me know if that would nt be the case from now.

Reviewer 4: Register is confusing to users. We're just updating the Environment property. We don't need another verb beyond update.

`rad env register recipe-pack recipe-pack-nam` --> `rad environment update myEnv --recipe-packs computeRecipePack`

This is described in the [feature spec](https://github.com/radius-project/design-notes/blob/main/features/2025-06-compute-extensibility-feature-spec.md#user-story-4b-as-a-developer-i-want-to-define-an-application-that-can-be-deployed-to-standard-aci-confidential-aci-and-kubernetes-compute-environments-without-changing-the-core-application-logic-so-that-i-can-maintain-a-consistent-application-definition-across-different-deployment-targets).

Author: 👍

6. Show environment details:

Based on whether the environment namespace is Applications.Core or Radius.Core, the outputs differ. Eventually Applications.Core support will be removed.

```
$ rad environment show my-env
RESOURCE            TYPE                      GROUP     STATE
my-env              Radius.Core/environments  default   Succeeded

> RECIPE PACK         RESOURCE TYPE                    RECIPE KIND     RECIPE VERSION      RECIPE LOCATION

Reviewer 4: What is Recipe Version and why do we have a column for something that is not designed or implemented? Can we remove for now?

computeRecipePack   Radius.Compute/containers        terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48
computeRecipePack   Radius.Security/secrets          terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/security/secrets?ref=v0.48
computeRecipePack   Radius.Storage/volumes           terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/storage/volumes?ref=v0.48
dataRecipePack      Radius.Data/redisCaches          terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/data/redisCaches?ref=v0.48
```

7. List recipes in environment:

```

> $ rad recipe list -environment my-env

Author: I think we woudl still want to use recipe packs to manage them, since it helps better organization, graph visibility, lifecycle and other benfets of modelling recipe packs as a resource seperate from environment and most likely in future theer would be more recipes to be added to the pack. 

It would be good to make it easier for the users to craete a recipe pack with a single recipe using a simple command though. 

perhaps  somethign like below?

rad recipe-pack create recipe-pack-name  --template-kind bicep --recipe-template-path template_path --recipe-resource-type Applications.Datastores/mongoDatabases --recipe-parameters throughput=400

> $ rad recipe list -environment my-env

Reviewer 4: All Recipes are specified in a Recipe Pack. That's the only option. It would be good to have an imperative command. I'll let Will comment on the priority of that since he didn't include it in the feature spec. 

If we did have a command it would look like:
```
rad recipe-pack create <recipe-pack-name> \
  --recipe-resource-type Applications.Datastores/mongoDatabases \
  --recipe-kind terraform \
  --recipe-location <recipe-location> \
  --recipe-parameters throughput=400
```

> $ rad recipe list -environment my-env

Author: for now, no shortcuts.

#### Registering recipes to multiple environments

As an operator I am responsible for creating Radius Environments using which developers can deploy their applications. As part of creating the environment, I manually link recipes one by one using `rad recipe register` or by updating the environment definition. I have 100s of environments which mostly use the same recipes. Piecing the same recipes together for each environment feels like rework.

#### Sharing Recipe Packs Across Environments and Organizations

As an operator, I want to share and reuse Recipe Packs across different environments or organizations. Instead of manually registering individual recipes, I can import a pre-bundled Recipe Pack (e.g., for Kubernetes or ACI) published by a provider or another team. This streamlines environment setup and reduces errors.

## Design

### Design Overview

In general, Radius.Core namespace has resources whose schema should be non-editable so that Radius can work as expected, for example Applications, Environments and recipePacks. These resources must be provisioned imperatively  and their schema must be protected. With this constraint in mind, 2 approaches are possible: 

> 1. Radius.Core/recipePacks provisioned imperatively by Applications RP

Author: recipe-packs dont have a recipe written in declarative languages.

> 2. Radius.Core/recipePacks provisioned Manually by Dynamic RP

Author: No, there was a possiblity to define a yaml schema for recipe pack and let dynamic rp provision it. But that will not work for the reasons covered in the doc ( custom dele/ update logics)

|---------|-----------| ------------------------------------------|
| **Tooling complexity** | ❌ Higher (TSP, converter, schema, API implemenetations needed) | ✅ Lower (YAML only, dynamic resource controllers reused)|
| **Versioning**  | ❌ requires versioning support in Radius  |✅ Supports schema versioning| 
| **Custom implementation for operations**  | ✅ can customize details of CRUDL|  ❌ falls back on dynamic resource controllers  |

Based on the above differences, *we choose Radius.Core/recipePacks to be provisioned imperatively by Applications RP as a first class Radius resource*. The main reason for this decision is that Radius Core resources such as enviroments and recipe-packs could have complex and custom deletion logic compared to what a dynamic resource deletion does. For instance, we need a cascade of deletion when an environment is deleted, or we might want to restrict deleting a recipe pack that is referenced in one or more environment. It is important to support robust core management operations of these resources where as the versioning of the type can follow once Radius versioning support is available. 

> We should make sure the rad resource-type commands cannot alter the schema of these types as part of schema validation (this namespace is reserved for Radius's use). Appropriate error message should be provided to the user. 

Reviewer 4: Since Environments and Applications do now have Recipes, they would not appear in Recipe Packs. @brooke-hamilton remember that Recipe Packs only have Recipes. The Resource Type definitions are separate.

### Other alternatives considered 

***Embed all recipe mappings inline in the Environment***

This is similar to what we have today. We could introduce a yaml spec similar to below, and when user executes a `rad recipe-pack register` this spec could be parsed and all recipes added to environment. 

```yaml
name: aci-production-pack
version: 1.0.0
description: "Recipe Pack for deploying to ACI in production."
recipes:
  - resourceType: "Radius.Compute/containers@2025-05-01-preview"   
    recipeKind: "bicep"

>     recipeLocation: "oci://ghcr.io/my-org/recipes/core/aci-container:1.2.0"

Reviewer 3: naming: recipeLocation versus recipeTemplatePath

      skuName: "premium"
```

Pros: 

- Most compatible to what we have in Radius today and hence the fastest approach. 
- Solves the requirement for bulk registering recipes using single command with a one time effort of creating the yaml manifest

Cons: 

- Environment still stays a bloated object. Environment resource houses a lot of other properties and we could potentially risk hitting the mechanical limits that apply to serializing objects. 
- Add Radius commands to publish recipe-packs, similar to what we have for recipes today. 
- A list of recipes could potentially be managed as a collection, including having its own rbac and appearance in app graph. The above approach does not allow for that possibility.

***Store a URL to a YAML manifest in the Environment***

We could fetch the yaml when needed, and use the available recipe.

`rad environment update my-env --recipe-packs aci-production-pack='git::https://github.com/my-org/recipe-packs.git//aci-production-pack.yaml?ref=1.0.0'`

Pros:

- Helps manage the size of environment resource
- Solves the requirement for bulk registering recipes using single command with a one time effort of creating the yaml manifest

Cons:

> - For each provisioning of resource, we make a call to registry to fetch the list to check if the recipe is available, and then a call to the specified recipe location to fetch it. We could cache the list and construct an in-memory recipe pack ephemeral object. But we still do not get the benefits of recipe pack as a first-class resource type. 

Reviewer 3: Not very clear how caching could work here or be different from pulling data from current store. Do you want to elaborate.

   
This flow should come in place if the environment used for deploying is of type Radius.Core/environments only. We retain current behavior for Applications.Core/environments.

As part of Radius.Core/environments design/implementation below points should be considered:

1. Finalize the `Radius.Core/environments` details and then

2. Add schema / swagger changes to support the `Radius.Core/environments` resource type

3. Add convertors for handling conversions from and to version agnostic data model. 

4. Add backend/controller support for creating/updating/deleting the resource. 

5. rad env register should support registering a recipe pack to Radius.Core/environments resource, and disallow recipes.

#### Dynamic RP changes

In Dynamic RP, while deploying a dynamic resource:

1. Add support to look up the `Radius.Core/environments` that is in use, fetch environment's recipe-pack ids

> 2. Go over the recipe packs registered in environment one by one until the first recipe pack holding the recipe for the resource type of interest is found.  

Reviewer 2: it should not be "the first recipe pack holding the recipe for the resource type of interest" but "the _only_ recipe pack holding the recipe for the resource type of interest" since the recipePacks and Environments resources should disallow duplicate recipes for the same resource types. See [user story 1](https://github.com/radius-project/design-notes/blob/main/features/2025-06-compute-extensibility-feature-spec.md#user-story-1-as-a-platform-engineer-i-want-to-create-a-recipe-pack-that-bundles-multiple-recipes-for-core-resource-types-so-that-i-can-easily-register-and-manage-them-in-a-radius-environment) from the feature spec for more details:

> Only one Recipe per Resource Type is allowed within a single Recipe Pack. If multiple Recipes for the same Resource Type are included, the creation process should fail with an appropriate error message.

> If Recipes are duplicated across the Recipe Packs added to an Environment, Radius will fail the environment deployment with a clear error message indicating the conflicting Recipes. This ensures that recipe conflicts are detected early and prevents unpredictable behavior. The duplication is detected at the Resource Type level, e.g. if both computeRecipePack and dataRecipePack contain a Recipe for the same Resource Type, the deployment will fail. Note that because we won't allow named Recipes in Recipe Packs, the conflict will only occur if the same Resource Type is specified in multiple Recipe Packs.

* Support `rad cli` commands that enable CRUDL operations on resources of type Radius.Core/recipePacks 
* Design and support Radius.Core/environment schema in Applications RP
* Support `rad cli` commands that enable managing Radius.Core/environment resources through CRUDL operations on this type of resource
* Support `rad cli` commands that enable registering recipe-packs to a `Radius.Core/environments` environment resource
* Enhance Dynamic RP, Applications RP and UCP to support the feature.

### Schema and API design

As part of supporting Recipe Pack as a resource type, at a high-level, we define a recipePacks.tsp
   
```tsp
namespace Applications.Core; 

@doc("The recipe pack resource") 
model RecipePackResource  
is TrackedResourceRequired<RecipePackProperties, "recipePacks"> { 
@doc("recipe pack name") 
@key("recipePackName") 
@path 
@segment("recipePacks") 
name: ResourceNameString; 
} 

@doc("Recipe Pack properties") 
model RecipePackProperties { 
@doc("The status of the asynchronous operation.") 
@visibility("read") 
provisioningState?: ProvisioningState; 

@doc("Description of what this recipe pack provides") 
description?: string; 

> 

Author: created https://github.com/radius-project/radius/issues/10831 to track this.

1. Show recipe pack details:

```
$ rad recipe-pack show computeRecipePack

RESOURCE                TYPE                      GROUP     
computeRecipePack       Radius.Core/recipePacks   default   

RESOURCE TYPE                    RECIPE KIND          RECIPE LOCATION    RECIPE PARAMETERS
Radius.Compute/containers        terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48  allowPlatformOptions(bool)
Radius.Security/secrets          terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/security/secrets?ref=v0.48
Radius.Storage/volumes           terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/storage/volumes?ref=v0.48
```

3. List recipe packs:

```
$ rad recipe-pack list [-g mygroup]

Lists all recipe packs in all scopes. 

RESOURCE                TYPE                      GROUP     STATE
computeRecipePack       Radius.Core/recipePacks   default   Succeeded
dataRecipePack          Radius.Core/recipePacks   default   Succeeded
```

> if -g is provided, list recipe packs in that scope.

Reviewer 4: ```suggestion
if --group or -g is provided, results are filtered for the specified Resource Group.
```

4. Show environment

```
$ rad environment show my-env
RESOURCE            TYPE                            GROUP     STATE
my-env              Radius.Core/environments        default   Succeeded

RECIPE PACKS        
computeRecipePack   
dataRecipePack      
```

5. Delete recipe pack:

```
$ rad recipe-pack delete <recipe-pack-name>
```

We could delete recipe-packs that are not referenced by any environment in any resource-group. If referenced by environments, the delete should also update the environment to not have the recipepack ID in its list of recipe packs. Allowance of recipe pack deletion when there are resources refering to it is in parity with today's recipe behavior. 

> By default, delete looks for recipe-pack in current group. It should take in optional scope to delete recipe packs in Other scope. 

Reviewer 4: ```suggestion
The delete command deletes the Recipe Pack and removes the Recipe Pack from all existing Environments. Because the Recipe is stored on each resource's status property, Radius retains details of the actual Recipe used to deploy the resource. Therefore, deleting a Recipe Pack only impacts the ability to deploy new resources using the referenced Recipes. This is in parity with today's Recipe behavior. 

By default, delete looks for Recipe Packs in the current Resource Group. Users can also specify the --group or -g parameter to reference a different Resource Group.
```
Rewrote for clarity, but no actual changes.

> 

Author: @willtsai  @zachcasper pls review. This is the behavior for recipe-packs, based on our decision to drop tagging support

* we are not supporting "scheme" (http|https|...). We can use the information in recipe location to determine that. We might have to introduce it back if we support other kinds of location for recipes in future.We might also have to introduce it for functional tests.
  
* The operations are all Synchronous, since Recipe Pack is a lightweight configuration resource.
  
* We maintain a reverse index into environment IDs so that we can handle CRUDL operations gracefully. For example, only a recipe pack that is not referenced by any environment can be deleted or updated. 

* We allow the users tp input a digest for a recipe to enhance security. More about this in [Verifying Recipy Integrity](#security)

  
#### Examples

Below is a sample bicep definition of a recipe pack resource:

```bicep
resource computeRecipePack 'Radius.Core/recipePacks@2026-01-01-preview' = {
  name: 'computeRecipePack'
  description: "Recipe Pack for deploying to Kubernetes."
  properties: {
    recipes: { 
      'Radius.Compute/containers': {
        recipeKind: 'terraform'
        recipeLocation: 'https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48'

>         recipeDigest: 'sha256:4g5h6i7j8k9l0m1n2o3p4q5r6s7t8u9v0w1x2y3z4a5b6c7d8e9f0g1h2i3j4k5'

Author: Digest is optional. I think Radius should be able to get the digest too, I will research on the details as part of the digest support  story.

1. Show recipe pack details:

```
$ rad recipe-pack show computeRecipePack............................................

RESOURCE                TYPE                      GROUP     STATE
computeRecipePack       Radius.Core/recipePacks   default   Succeeded

RESOURCE TYPE                    RECIPE KIND          RECIPE LOCATION    RECIPE PARAMETERS
Radius.Compute/containers        terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/compute/containers/kubernetes?ref=v0.48  allowPlatformOptions(bool)
Radius.Security/secrets          terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/security/secrets?ref=v0.48
Radius.Storage/volumes           terraform                           https://github.com/project-radius/resource-types-contrib.git//recipes/storage/volumes?ref=v0.48
```

3. List recipe packs:

```
$ rad recipe-pack list [-g mygroup]

Lists all recipe packs in all scopes. 

RESOURCE                TYPE                      GROUP     STATE
computeRecipePack       Radius.Core/recipePacks   default   Succeeded

> dataRecipePack          Radius.Core/recipePacks   default   Succeeded

Reviewer 4: Can we NOT show the state for a non-deployed resource like RecipePacks and TerraformSettings?

Author: remove state

$ rad environment create myEnv \
  --recipe-packs computeRecipePack
This command creates an environment with the `recipePacks[]` property populated with `computeRecipePack`.

$ rad environment update myEnv \
  --recipe-packs computeRecipePack
The command should use the resource ID to identify the environment is `Radius.Core\environment` resource and only then allow the `recipePack` property to be updated. If the namespace is `Applications.Core`, an error is provided.

$ rad environment update myEnv \
  --recipe-packs otherResourceGroup/computeRecipePack
If the Recipe Pack is in a different Resource Group than the Environment, the Resource Group is passed as a prefix to the Recipe Pack name.
```

1. Show environment details:

Based on whether the environment namespace is Applications.Core or Radius.Core, the outputs differ. Eventually Applications.Core support will be removed.

```
$ rad environment show my-env
RESOURCE            TYPE                      GROUP     STATE
my-env              Radius.Core/environments  default   Succeeded

RECIPE PACK        
computeRecipePack  
dataRecipePack
```

7. List recipes in environment:

```
$ rad recipe list -environment my-env

> RECIPE PACK             RESOURCE TYPE                    RECIPE KIND     RECIPE VERSION      RECIPE LOCATION

Reviewer 4: Can we remove the empty RECIPE VERSION column? If/when we ever implement a version on a recipe we can add the column back.

### Graph support

> Recipe packs will not be displayed in application graphs since the are operatiors concept and not part of an application as a component.

Reviewer: We should add a section to the dashboard to display the recipe packs.

Author: the dashboard should be able to list recipe packs as well as what it captures today for recipes.
